第14章单测运行失败

第14章的单元测试在Ubuntu、Windows和macOS三个环境下都运行失败：<https://github.com/ZZy979/Beginning-Python-code/actions/runs/8503483952>

但是三个环境下的报错信息又不完全相同。

* Ubuntu：有 "AssertionError: 5 != 0" 和 "OSError: [Errno 98] Address already in use" 两种错误。前者是因为**未获取到服务器程序的标准输出**；后者是因为在UNIX系统上，服务器程序终止后需要等待一段时间才能再次使用同一个端口号，见[socket模块](https://docs.python.org/3/library/socket.html#example) "Example" 一节最后的说明。

```
======================================================================
FAIL: test_forking_server (ch14.test.Ch14Tests.test_forking_server)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 33, in test_forking_server
    self._test_socket_server('forking_server.py')
  File "/home/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 21, in _test_socket_server
    self.assertEqual(len(client_results), stdout.count('Got connection from'))
AssertionError: 5 != 0

======================================================================
FAIL: test_minimal_server (ch14.test.Ch14Tests.test_minimal_server)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/home/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 26, in test_minimal_server
    self._test_socket_server('minimal_server.py')
  File "/home/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 20, in _test_socket_server
    self.assertFalse(stderr)
AssertionError: 'Traceback (most recent call last):\n  File "/home/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/minimal_server.py", line 7, in <module>\n    s.bind((host, port))\nOSError: [Errno 98] Address already in use\n' is not false
```

* Windows：只有 "AssertionError: 5 != 0" 一种错误。
* macOS：有 "AssertionError: 5 != 0" 和 "socket.gaierror: [Errno 8] nodename nor servname provided, or not known" 两种错误。后者是因为/etc/hosts文件中没有将主机名映射到IP地址127.0.0.1。

```
======================================================================
FAIL: test_forking_server (ch14.test.Ch14Tests.test_forking_server)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 33, in test_forking_server
    self._test_socket_server('forking_server.py')
  File "/Users/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 21, in _test_socket_server
    self.assertEqual(len(client_results), stdout.count('Got connection from'))
AssertionError: 5 != 0

======================================================================
FAIL: test_minimal_server (ch14.test.Ch14Tests.test_minimal_server)
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/Users/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 26, in test_minimal_server
    self._test_socket_server('minimal_server.py')
  File "/Users/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/test.py", line 20, in _test_socket_server
    self.assertFalse(stderr)
AssertionError: 'Traceback (most recent call last):\n  File "/Users/runner/work/Beginning-Python-code/Beginning-Python-code/ch14/minimal_server.py", line 7, in <module>\n    s.bind((host, port))\nsocket.gaierror: [Errno 8] nodename nor servname provided, or not known\n' is not false
```

在个人Windows计算机上，使用PyCharm运行单测能够通过，在命令行运行则失败。另外，在PyCharm的Python Console执行服务器脚本能够获取到标准输出：

```python
>>> import subprocess
>>> proc = subprocess.Popen(['python', 'ch14/minimal_server.py'], stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding='utf-8', text=True)
# 在其他命令行窗口执行python minimal_client.py
>>> proc.kill()
>>> proc.communicate()
("Got connection from ('192.168.47.1', 5228)\nGot connection from ('192.168.47.1', 5229)\nGot connection from ('192.168.47.1', 5230)\nGot connection from ('192.168.47.1', 5235)\n", '')
```

而在命令行运行就获取不到：

```python
>>> import subprocess
>>> proc = subprocess.Popen(['python', 'ch14/minimal_server.py'], stdout=subprocess.PIPE, stderr=subprocess.PIPE, encoding='utf-8', text=True)
# 在其他命令行窗口执行python minimal_client.py
>>> proc.kill()
>>> proc.communicate()
('', '')
```

--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -24,4 +24,7 @@ jobs:
           python -m pip install --upgrade pip
           pip install -r requirements.txt
       - name: Run test
+        env:
+          PYTHONUNBUFFERED: 1
+          PYTHONIOENCODING: UTF-8
         run: python -m unittest -v

--- a/README.md
+++ b/README.md
@@ -103,6 +103,7 @@ python -m unittest
 * [代码清单14-4 分叉服务器](ch14/forking_server.py)
 * [代码清单14-5 线程化服务器](ch14/threading_server.py)
 * [代码清单14-6 使用select的简单服务器](ch14/select_server.py)
+* [Telnet客户端](ch14/telnet_client.py)
 * [代码清单14-7 使用poll的简单服务器](ch14/poll_server.py)
 * [代码清单14-8 使用Twisted的简单服务器](ch14/twisted_server.py)
 * [代码清单14-9 使用LineReceiver改进的日志服务器](ch14/twisted_line_server.py)

--- a/ch14/forking_server.py
+++ b/ch14/forking_server.py
@@ -8,5 +8,6 @@ class Handler(StreamRequestHandler):
         self.wfile.write(b'Thank you for connecting')
 
 
+ForkingTCPServer.allow_reuse_address = True
 server = ForkingTCPServer(('', 1234), Handler)
 server.serve_forever()

--- a/ch14/minimal_client.py
+++ b/ch14/minimal_client.py
@@ -2,7 +2,7 @@ import socket
 
 s = socket.socket()
 
-host = socket.gethostname()
+host = 'localhost'
 port = 1234
 
 s.connect((host, port))

--- a/ch14/minimal_server.py
+++ b/ch14/minimal_server.py
@@ -1,8 +1,9 @@
 import socket
 
 s = socket.socket()
+s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
 
-host = socket.gethostname()
+host = ''
 port = 1234
 s.bind((host, port))
 

--- a/ch14/poll_server.py
+++ b/ch14/poll_server.py
@@ -16,7 +16,7 @@ p.register(s)
 while True:
     events = p.poll()
     for fd, event in events:
-        if fd in fdmap:
+        if fd == s.fileno():
             c, addr = s.accept()
             print('Got connection from', addr)
             p.register(c)

--- a/ch14/poll_server.py
+++ b/ch14/poll_server.py
@@ -2,8 +2,9 @@ import select
 import socket
 
 s = socket.socket()
+s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
 
-host = socket.gethostname()
+host = ''
 port = 1234
 s.bind((host, port))
 

--- a/ch14/select_server.py
+++ b/ch14/select_server.py
@@ -2,8 +2,9 @@ import select
 import socket
 
 s = socket.socket()
+s.setsockopt(socket.SOL_SOCKET, socket.SO_REUSEADDR, 1)
 
-host = socket.gethostname()
+host = ''
 port = 1234
 s.bind((host, port))
 

--- a/ch14/socketserver_minimal_server.py
+++ b/ch14/socketserver_minimal_server.py
@@ -8,5 +8,6 @@ class Handler(StreamRequestHandler):
         self.wfile.write(b'Thank you for connecting')
 
 
+TCPServer.allow_reuse_address = True
 server = TCPServer(('', 1234), Handler)
 server.serve_forever()

--- a/ch14/telnet_client.py
+++ b/ch14/telnet_client.py
@@ -1,6 +1,5 @@
-import socket
 import sys
 from telnetlib import Telnet
 
-with Telnet(socket.gethostname(), 1234) as tn:
+with Telnet('localhost', 1234) as tn:
     tn.write(sys.stdin.read().encode())

--- a/ch14/test.py
+++ b/ch14/test.py
@@ -9,14 +9,12 @@ class Ch14Tests(test_utils.TestCase):
     dirname = 'ch14'
 
     def _test_socket_server(self, server_script):
-        with self.popen(server_script) as server_proc:
+        def _call_clients():
             with ThreadPoolExecutor() as executor:
                 client_futures = [executor.submit(self.run_script, 'minimal_client.py') for _ in range(5)]
-                client_results = [f.result() for f in client_futures]
-            server_proc.kill()
-            stderr = server_proc.stderr.read()
-            stdout = server_proc.stdout.read()
+                return [f.result() for f in client_futures]
 
+        stdout, stderr, client_results = self.run_server(server_script, client_func=_call_clients)
         self.assertFalse(stderr)
         self.assertEqual(len(client_results), stdout.count('Got connection from'))
         for r in client_results:
@@ -37,15 +35,14 @@ class Ch14Tests(test_utils.TestCase):
 
     def _test_logger_server(self, server_script):
         data = [f'client{i}\n' for i in range(5)]
-        with self.popen(server_script) as server_proc:
+
+        def _call_clients():
             with ThreadPoolExecutor() as executor:
                 for d in data:
                     executor.submit(self.run_script, 'telnet_client.py', input=d)
-            server_proc.kill()
-            stderr = server_proc.stderr.read()
-            stdout = server_proc.stdout.read()
 
-        self.assertFalse('', stderr)
+        stdout, stderr, _ = self.run_server(server_script, client_func=_call_clients)
+        self.assertFalse(stderr)
         self.assertEqual(len(data), stdout.count('Got connection from'))
         self.assertEqual(len(data), stdout.count('disconnected'))
         for d in data:

--- a/ch14/threading_server.py
+++ b/ch14/threading_server.py
@@ -8,5 +8,6 @@ class Handler(StreamRequestHandler):
         self.wfile.write(b'Thank you for connecting')
 
 
+ThreadingTCPServer.allow_reuse_address = True
 server = ThreadingTCPServer(('', 1234), Handler)
 server.serve_forever()

--- a/test_utils.py
+++ b/test_utils.py
@@ -2,6 +2,7 @@ import pathlib
 import shlex
 import subprocess
 import sys
+import time
 import unittest
 
 
@@ -50,12 +51,25 @@ class TestCase(unittest.TestCase):
         actual_output = self.run_script(script, args, input, input_file, cwd).stdout
         self.assertEqual(expected_output, actual_output)
 
-    def popen(self, script, args='', cwd=None):
-        """运行指定的脚本，返回subprocess.Popen对象。"""
-        cmd = [sys.executable, script, *shlex.split(args)]
-        return subprocess.Popen(
+    def run_server(self, server_script, server_args='', wait_time=1, client_func=None, cwd=None):
+        """运行指定的服务器脚本，之后调用客户端，并返回服务器的输出。
+
+        :param server_script: str 服务器脚本文件名，相对于cwd参数指定的目录
+        :param server_args: str 服务器命令行参数，空格分隔
+        :param wait_time: float 启动服务器后的等待时间，单位：秒
+        :param client_func: Callable[[], Any] 调用客户端的函数
+        :param cwd: str 当前工作目录，默认为self.dir
+        :return: (str, str, Any) 服务器的标准输出、标准错误和客户端函数的返回结果
+        """
+        cmd = [sys.executable, server_script, *shlex.split(server_args)]
+        server_proc = subprocess.Popen(
             cmd, stdin=subprocess.DEVNULL, stdout=subprocess.PIPE, stderr=subprocess.PIPE,
             cwd=self.get_cwd(cwd), encoding='utf-8', text=True)
+        time.sleep(wait_time)
+        client_results = client_func() if callable(client_func) else None
+        server_proc.kill()
+        stdout, stderr = server_proc.communicate()
+        return stdout, stderr, client_results
 
     def get_cwd(self, cwd=None):
         return cwd if cwd is not None else self.dir

